packages/babel-plugin-proposal-discard-binding/src/utils.ts

```diff
@@ -0,0 +1,165 @@
+        parentPath.scope.registerBinding(bindingKind, path, childPath);
+        return;
+      }
+      case "AssignmentExpression":
+      case "ForInStatement":
+      case "ForOfStatement":
+        parentPath.scope.push({ id: cloneNode(path.node) });
+        return;
+      case "CatchClause":
+        parentPath.scope.registerBinding("let", parentPath);
+        return;
+      default:
+        if (parentPath.isFunction()) {
+          parentPath.scope.registerBinding("param", path, childPath);
+          return;
+        }
+    }
+    childPath = parentPath;
+  }
+  /* c8 ignore next */
+  throw new Error("@babel/plugin-proposal-discard-binding: Internal Error");
+}
+
+export function transformVoidPattern(
+  path: NodePath<t.VoidPattern>,
+): NodePath<t.Identifier> | null {
+  if (path.parentPath.isArrayPattern()) {
+    return transformVoidPatternIntoNull(path);
+  } else {
+    return transformVoidPatternIntoUidIdentifier(path);
+  }
+}
+
+function* iterateVoidPatternsInLVal(
+  path: NodePath<t.LVal | t.PatternLike>,
+): Generator<NodePath<t.VoidPattern>> {
+  switch (path.type) {
+    case "ArrayPattern":
+      for (const elementPath of path.get("elements")) {
+        yield* iterateVoidPatternsInLVal(elementPath);
+      }
+      break;
+    case "ObjectPattern":
+      for (const propertyPath of path.get("properties")) {
+        if (propertyPath.isObjectProperty()) {
+          yield* iterateVoidPatternsInLVal(
+            propertyPath.get("value") as NodePath<t.Pattern>,
+          );
+        }
+      }
+      break;
+    case "AssignmentPattern":
+      yield* iterateVoidPatternsInLVal(path.get("left"));
+      break;
+    case "VoidPattern":
+      yield path;
+      break;
+    default:
+      break;
+  }
+}
+
+function* iterateVoidPatternInVariableDeclaration(
+  path: NodePath<t.VariableDeclaration>,
+): Generator<NodePath<t.VoidPattern>> {
+  for (const declaratorPath of path.get("declarations")) {
+    yield* iterateVoidPatternsInLVal(declaratorPath.get("id"));
+  }
+}
+
+export function hasVoidPatternInVariableDeclaration(
+  path: NodePath<t.VariableDeclaration>,
+): boolean {
+  return !iterateVoidPatternInVariableDeclaration(path).next().done;
+}
+
+export function transformVoidPatternInLVal(
+  path: NodePath<t.LVal | t.PatternLike>,
+) {
+  for (const voidPatternPath of iterateVoidPatternsInLVal(path)) {
+    transformVoidPattern(voidPatternPath);
+  }
+}
+
+export function removeTrailingVoidPatternsFromParams(
+  path: NodePath<t.Function>,
+) {
+  const paramPaths = path.get("params");
+  for (let i = paramPaths.length - 1; i >= 0; i--) {
+    const paramPath = paramPaths[i];
+    if (paramPath.isVoidPattern()) {
+      paramPath.remove();
+    } else {
+      break;
+    }
+  }
+}
+
+// https://tc39.es/ecma262/#sec-isanonymousfunctiondefinition
+// We don't test anonymous function / arrow function because they must not be disposable
+export function isAnonymousFunctionDefinition(
```

Comment:
I don't think we need this check. There is no good reason for anybody to write `using void = () => {}`, but if they do we might as well inject (`0,`) and avoid having two different code paths for classes and functions.

Reply:
Well, if they do they will get a runtime error for sure. Added the function / arrow function check.